### ISSUE TITLE
reference is not canonical

### ISSUE THREAD

**Issue by goern** [xeipuuv/gojsonschema#92]
Hello,
 I have used the example code to validate some JSON, it gives me an error `--- FAIL: TestValidate (2.10s)
	validator_test.go:28: Error (Reference {0xc8201f4680 {[]} false false false true false} must be canonical)` which I can't understand. I am using the schema at http://goern.github.io/grasshopper/nulecule/spec/0.0.2/schema.json

**Comment by xeipuuv** [comment]
Hard to tell, please post the source code, and the error stack.
You mentioned using the example code, are you using a URI "file://" ? 
If so, it could be the issue and try with "http://" URI.

**Comment by goern** [comment]
heya, thanks for that super fast reply!

The error isnt much, its a go test:
```
=== RUN   TestValidate
--- FAIL: TestValidate (0.39s)
	validator_test.go:28: Error (Reference {0xc8201b0d80 {[]} false false false true false} must be canonical)
FAIL
FAIL	github.com/goern/grasshopper/nulecule/utils	0.398s
```

The [test itself](https://github.com/goern/grasshopper/blob/feature/nulecule-validate/nulecule/utils/validator_test.go)  and [the source](https://github.com/goern/grasshopper/blob/feature/nulecule-validate/nulecule/utils/validator.go#L37). I tried with https and http and local file, only the position of the true within the error message changes.

**Comment by xeipuuv** [comment]
There is a file reference to param.json in the schema http://goern.github.io/grasshopper/nulecule/spec/0.0.2/graph.json

**Comment by goern** [comment]
it still fails with the same error, I do only use http urls for schema and document, and I still use the same src as mentioned earlier.

pls re-open

**Comment by xeipuuv** [comment]
in http://goern.github.io/grasshopper/nulecule/spec/0.0.2/param.json

"$ref": "file:constraint.json"

**Comment by goern** [comment]
```
=== RUN   TestValidate
--- PASS: TestValidate (2.10s)
```

I owe you two drinks, sorry for that!

**Comment by xeipuuv** [comment]
2 drinks ? well, go for a bottle then haha.

**Comment by goern** [comment]
let me know when you travel in western germany ;)

**Comment by xeipuuv** [comment]
I used to live in Strasboug, France, very very close.
Glad it works.

**Comment by baolinCloud** [comment]
@xeipuuv , hello,  I met the same problem. 
Code listed below:
`schemaLoader := gojsonschema.NewReferenceLoader("file:///d/metricSchema.json")`
`documentLoader := gojsonschema.NewReferenceLoader("file:///d//testMetric.json")`

And the error:
`panic: Reference {0xc042046c00 {[]} false true false true false} must be canonical`

And I find that you said before,

> Hard to tell, please post the source code, and the error stack.
You mentioned using the example code, are you using a URI "file://" ?
If so, it could be the issue and try with "http://" URI.

NewReferenceLoader with file does not work?

**Comment by xeipuuv** [comment]
Seems like there must be a non canonical reference inside the metricSchema.json itself ("$ref")

**Comment by baolinCloud** [comment]
Maybe it does not work in windows. I will try to use NewStringLoader.

**Comment by bearx3f** [comment]
in windows,

`file:///C:/path/to/your/schema/schema.json`

**Comment by ss22ever** [comment]
@xeipuuv thanks for the code, I also get similar error -

`panic: Reference {0xc042114300 {[]} false true false true false} must be canonical

goroutine 1 [running]:
main.main()
        C:/resume/src/validator.go:17 +0x45c
exit status 2`

following is the gist to the corresponding files:
https://gist.github.com/ss22ever/9979650e41ccde523c2df419d451dd42

**Comment by codepushr** [comment]
Actually this error also appears when the path is invalid. This should not be the case IMHO.

**Comment by FrancisBilla** [comment]
How do you a relative path ?

**Comment by mattwelke** [comment]
Ran into this today trying a small sample from the main website, except I wanted to use relative paths instead.

![image](https://user-images.githubusercontent.com/7719209/127227597-94151297-41f7-44cb-a2b9-64ec8b85cc08.png)

(document.json and schema.json have the contents from the main page)

```go
func main() {
	schemaLoader := gojsonschema.NewReferenceLoader("file://schema.json")
	documentLoader := gojsonschema.NewReferenceLoader("file://document.json")

	result, err := gojsonschema.Validate(schemaLoader, documentLoader)
	if err != nil {
		panic(err.Error())
	}

	if result.Valid() {
		fmt.Printf("The document is valid\n")
	} else {
		fmt.Printf("The document is not valid. see errors :\n")
		for _, desc := range result.Errors() {
			fmt.Printf("- %s\n", desc)
		}
	}
}
```

```
~/go-jsonschema-test > go run main.go 
panic: Reference file://schema.json must be canonical

goroutine 1 [running]:
main.main()
        /home/matt/go-jsonschema-test/main.go:15 +0x2c7
exit status 2
```

**Comment by mattwelke** [comment]
Fixed by changing the code to:

```go
schemaLoader := gojsonschema.NewReferenceLoader("file://./schema.json")
documentLoader := gojsonschema.NewReferenceLoader("file://./document.json")
```

I found https://github.com/xeipuuv/gojsonschema/issues/140 which happened to show using a relative path that worked. It looks like for this library, even though the paths "schema.json" and "document.json" are valid relative paths for a shell, they aren't valid for the Go process loading the files via this library. Not sure why.
